agent/agent.go

```diff
@@ -0,0 +1,825 @@
+	heartbeatTimeoutCh chan struct{}
+
+	opts    Options
+	logger  xlog.Logger
+	metrics *opAgentMetrics
+	doneCh  chan struct{}
+	closeCh chan struct{}
+}
+
+type newProcessMonitorFn func(exec.Cmd, exec.ProcessListener) (exec.ProcessMonitor, error)
+
+// New creates and retuns a new Operator Agent
+func New(
+	opts Options,
+) (Agent, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	if err := canaryWriteTest(opts.WorkingDirectory()); err != nil {
+		return nil, err
+	}
+
+	agent := &opAgent{
+		opts:                opts,
+		logger:              opts.InstrumentOptions().Logger(),
+		metrics:             newAgentMetrics(opts.InstrumentOptions().MetricsScope()),
+		newProcessMonitorFn: exec.NewProcessMonitor,
+		doneCh:              make(chan struct{}),
+		closeCh:             make(chan struct{}),
+	}
+	go agent.reportMetrics()
+	return agent, nil
+}
+
+func (o *opAgent) Close() error {
+	o.closeCh <- struct{}{}
+	<-o.doneCh
+	return nil
+}
+
+func canaryWriteTest(dir string) error {
+	fi, err := os.Stat(dir)
+	if err != nil {
+		return fmt.Errorf("unable to stat directory, [ err = %v ]", err)
+	}
+	if !fi.IsDir() {
+		return fmt.Errorf("path is not a directory")
+	}
+
+	fd, err := ioutil.TempFile(dir, defaultTestCanaryPrefix)
+	if err != nil {
+		return fmt.Errorf("unable to create canary file, [ err = %v ]", err)
+	}
+	os.Remove(fd.Name())
+
+	return nil
+}
+
+func updateBoolGauge(b bool, m tally.Gauge) {
+	if b {
+		m.Update(1)
+	} else {
+		m.Update(0)
+	}
+}
+
+func (o *opAgent) reportMetrics() {
+	reportTicker := time.NewTicker(defaultReportInterval)
+	for {
+		select {
+		case <-reportTicker.C:
+			running, exec, conf := o.state()
+			updateBoolGauge(running, o.metrics.running)
+			updateBoolGauge(exec != "", o.metrics.execTransferred)
+			updateBoolGauge(conf != "", o.metrics.confTransferred)
+		case <-o.closeCh:
+			reportTicker.Stop()
+			o.doneCh <- struct{}{}
+			return
+		}
+	}
+}
+
+func (o *opAgent) Running() bool {
+	return atomic.LoadInt32(&o.running) == 1
+}
+
+func (o *opAgent) state() (bool, string, string) {
+	o.RLock()
+	defer o.RUnlock()
+	return o.Running(), o.executablePath, o.configPath
+}
+
+func (o *opAgent) Start(ctx context.Context, request *m3em.StartRequest) (*m3em.StartResponse, error) {
+	o.logger.Infof("received Start()")
+	o.Lock()
+	defer o.Unlock()
+
+	if o.Running() {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "already running")
+	}
+
+	if o.executablePath == "" {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "agent missing build")
+	}
+
+	if o.configPath == "" {
+		return nil, grpc.Errorf(codes.FailedPrecondition, "agent missing config")
+	}
+
+	if err := o.startWithLock(); err != nil {
+		return nil, grpc.Errorf(codes.Internal, "unable to start: %v", err)
+	}
+
+	return &m3em.StartResponse{}, nil
+}
+
+func (o *opAgent) newListenerFn() func(error) {
+	return func(err error) {
+		if err == nil {
+			err = fmt.Errorf("test process terminated without error")
```

Comment:
I'm assuming all the processes we'd run under this tool are long running, they should not exit unless we explicitly `Stop()` them. So I treated clean exit of the process as an error too.